shipcat_cli/src/apply.rs

```diff
@@ -0,0 +1,317 @@
+/// It is provided for webhooks and notifications to make good interfaces.
+/// TODO: remove fork in helm module
+#[derive(Clone)]
+pub struct UpgradeInfo {
+    /// Name of service
+    pub name: String,
+    /// Metadata for service
+    pub metadata: Metadata,
+    /// Validated version string
+    pub version: String,
+    /// Chart the service is using
+    chart: String,
+    /// Validated region requested for installation
+    pub region: String,
+    /// Validated namespace inferred from region
+    pub namespace: String,
+    /// Computed diff string (if available)
+    pub diff: Option<String>,
+}
+
+impl UpgradeInfo {
+    /// Export the information needed for book-keeping
+    ///
+    /// Requires a valid manifest with a pre-determined version
+    pub fn new(mf: &Manifest) ->  Self {
+        UpgradeInfo {
+            name: mf.name.clone(),
+            version: mf.version.clone().expect("version must be set before calling UpgradeInfo::new"),
+            metadata: mf.metadata.clone().expect("metadata must exist on every manifest"),
+            chart: mf.chart.clone().expect("must have a chart for a manifest"),
+            region: mf.region.clone(),
+            namespace: mf.namespace.clone(),
+            diff: None,
+        }
+    }
+}
+
+
+
+// TODO: helm migration:
+// 1. make a simplified UpgradeMode (no need for most options)
+// 2. fallback way to get version in rolling once they are in all crds
+
+/// shipcat apply
+///
+/// This is the main entrypoint for cli upgrades of a service in a region.
+/// It currently uses tiller, but this function will be updated to bypass that soon.
+///
+/// The design goals of this is to support:
+/// - shipcat apply mysvc
+/// - shipcat apply mysvc -t semver
+/// In both non-rolling environments (preprod/prod), and rolling envs (complement).
+///
+/// To allow version-less apply in rolling environment, we have to verify with kube
+/// what the current version is.
+///
+/// The preparatory sanity
+///
+/// This is entry point should be used by both the cli to apply, and reconcile.
+/// It is also entirely responsible for sending webhooks on errors / successes.
+/// As such, it's entirely responsible for not propagating random errors here with `?`
+/// Every error cases is something that might need to be notified.
+pub fn apply(mfbase: Manifest,
+             force: bool,
+             region: &Region,
+             conf: &Config,
+             wait: bool,
+             override_version: Option<String>) -> Result<Option<UpgradeInfo>>
+{
+    if let Err(e) = webhooks::ensure_requirements(&region) {
+        warn!("Could not ensure webhook requirements: {}", e);
+    }
+    let svc = mfbase.name.clone();
+    // Internal sanity
+    assert_eq!(mfbase.region, region.name);
+    assert_eq!(mfbase.namespace, region.namespace);
+
+    // A version can be explicit in two ways: passed via -t SOMEVER, or set in manifest
+    let explicit_version = override_version.or(mfbase.version.clone());
```

Comment:
This behaviour already introduces some weird bugs. e.g., if you override a fixed version, it'll get rolled back on the next reconcile.

IMO, if a version is set in the manifest, it shouldn't be overridable:

```suggestion
    let explicit_version = mfbase.version.clone().or(override_version);
```

Reply:
Ah that's true. I'll put in a bail! for that.

Reply:
Thanks.

It's really confused me in the past, where I've accidentally hardcoded the version for dev and my deploys keep getting magically rolled back.